pkg/server/endpoints/authorized_entry_fetcher.go

```diff
@@ -0,0 +1,173 @@
+package endpoints
+
+import (
+	"context"
+	"time"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/pkg/server/authorizedentries"
+	"github.com/spiffe/spire/pkg/server/datastore"
+)
+
+var _ api.AuthorizedEntryFetcher = (*AuthorizedEntryFetcherWithEventsBasedCache)(nil)
+
+type AuthorizedEntryFetcherWithEventsBasedCache struct {
+	cache                *authorizedentries.Cache
+	clk                  clock.Clock
+	log                  logrus.FieldLogger
+	ds                   datastore.DataStore
+	cacheReloadInterval  time.Duration
+	pruneEventsOlderThan time.Duration
+	lastEventID          uint
+}
+
+func NewAuthorizedEntryFetcherWithEventsBasedCache(ctx context.Context, log logrus.FieldLogger, clk clock.Clock, ds datastore.DataStore, cacheReloadInterval, pruneEventsOlderThan time.Duration) (*AuthorizedEntryFetcherWithEventsBasedCache, error) {
+	log.Info("Building in-memory entry cache")
+	cache, err := buildCache(ctx, ds)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Completed building in-memory entry cache")
+
+	return &AuthorizedEntryFetcherWithEventsBasedCache{
+		cache:                cache,
+		clk:                  clk,
+		log:                  log,
+		ds:                   ds,
+		cacheReloadInterval:  cacheReloadInterval,
+		pruneEventsOlderThan: pruneEventsOlderThan,
+	}, nil
+}
+
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) FetchAuthorizedEntries(_ context.Context, agentID spiffeid.ID) ([]*types.Entry, error) {
+	return a.cache.GetAuthorizedEntries(agentID), nil
+}
+
+// RunUpdateCacheTask starts a ticker which rebuilds the in-memory entry cache.
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) RunUpdateCacheTask(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			a.log.Debug("Stopping in-memory entry cache hydrator")
+			return nil
+		case <-a.clk.After(a.cacheReloadInterval):
+			err := a.updateCache(ctx)
+			if err != nil {
+				a.log.WithError(err).Error("Failed to update entry cache")
+			}
+		}
+	}
+}
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) PruneEventsTask(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			a.log.Debug("Stopping event pruner")
+			return nil
+		case <-a.clk.After(a.pruneEventsOlderThan / 2):
+			a.log.Debug("Pruning events")
+			if err := a.pruneEvents(ctx, a.pruneEventsOlderThan); err != nil {
+				a.log.WithError(err).Error("Failed to prune events")
+			}
+		}
+	}
+}
+
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) pruneEvents(ctx context.Context, olderThan time.Duration) error {
+	if err := a.ds.PruneRegistrationEntriesEvents(ctx, olderThan); err != nil {
+		return err
+	}
+	return a.ds.PruneAttestedNodesEvents(ctx, olderThan)
```

Comment:
should we opportunistically try and prune agents even if entry pruning failed?

Reply:
Yea i thought about that too. The error handling would be more complicated since 1 or both could get error.

Reply:
it's not too bad if you capture each err in it's own variable and then just return errors.Join(err1, err2) or something

Reply:
updated to use errors.join. Good suggestion.